source/base/qprojector.cc

```diff
@@ -336,6 +336,134 @@ namespace internal
+                 reference_cell == ReferenceCells::Quadrilateral,
+               ExcInternalError());
+
+        const unsigned int n_faces = reference_cell.n_faces();
+        const unsigned int num_subfaces_or_face =
+          project_to_subface ?
+            reference_cell.face_reference_cell(0).n_isotropic_children() :
+            1;
+
+        unsigned int n_points = 0;
+        if (quadrature.size() == 1)
+          n_points = quadrature[0].size() * n_faces;
+        else
+          {
+            AssertDimension(quadrature.size(), n_faces);
+            for (const auto &q : quadrature)
+              n_points += q.size();
+          }
+
+        // new (projected) quadrature points and weights
+        std::vector<Point<2>> q_points;
+        q_points.reserve(2 * n_points * num_subfaces_or_face);
+        std::vector<double> weights;
+        weights.reserve(2 * n_points * num_subfaces_or_face);
+
+        auto process_face_or_subface =
+          [&](const unsigned int                 face,
+              const types::geometric_orientation orientation,
+              const unsigned int                 subface) {
+            const unsigned int quadrature_index =
+              quadrature.size() == 1 ? 0 : face;
+            const unsigned int n_q_points = quadrature[quadrature_index].size();
+
+            // handle the points
+            std::vector<Point<2>> points(n_q_points);
+
+            // orientation numbers::default_geometric_orientation is the
+            // standard orientation so invert lines for orientation
+            // numbers::reverse_line_orientation
+            if (project_to_subface)
+              {
+                const unsigned int local_subface =
+                  orientation == numbers::reverse_line_orientation ?
+                    num_subfaces_or_face - 1 - subface :
+                    subface;
+
+                dealii::QProjector<2>::project_to_subface(
+                  reference_cell,
+                  quadrature[quadrature_index],
+                  face,
+                  local_subface,
+                  points);
+              }
+            else
+              {
+                dealii::QProjector<2>::project_to_face(
+                  reference_cell, quadrature[quadrature_index], face, points);
+              }
+
+            if (orientation == numbers::reverse_line_orientation)
+              std::reverse(points.begin(), points.end());
+
+            std::copy(points.begin(),
+                      points.end(),
+                      std::back_inserter(q_points));
+
+
+            // handle the weights
+            std::vector<double> scaled_weights =
+              quadrature[quadrature_index].get_weights();
+
+            // triangular face 1 is given by (1,0) (0,1) so scale the
+            // weights by the length
+            if (reference_cell == ReferenceCells::Triangle && face == 1)
+              for (auto &w : scaled_weights)
+                w *= std::sqrt(2);
```

Comment:
Similarly, in #18306 we added `ReferenceCell::face_measure()` which better enables reference cell and dimension-independent programming.